cluster/main.tf

```diff
@@ -0,0 +1,112 @@
+  role_arn           = "arn:aws:iam::${local.account_id}:role/doitintl_eks_import_${local.region}"
+  cluster_issuer_url = replace(var.cluster_oidc_issuer_url, "https://", "")
+  doit_webhook_url   = "https://dev-app.doit.com/webhooks/v1/eks-metrics"
+}
+
+// Policy to allow the EKS Lens pod to interact with the S3 bucket
+data "aws_iam_policy_document" "doit_eks_lens_collector" {
+  count = var.ec2_cluster ? 0 : 1
+
+  statement {
+    effect = "Allow"
+
+    principals {
+      type = "Federated"
+      identifiers = [
+        "arn:aws:iam::${local.account_id}:oidc-provider/${local.cluster_issuer_url}",
+      ]
+    }
+
+    actions = [
+      "sts:AssumeRoleWithWebIdentity",
+    ]
+
+    condition {
+      test = "StringEquals"
+      values = [
+        "system:serviceaccount:doit-eks-metrics:doit-collector",
+      ]
+      variable = "${local.cluster_issuer_url}:sub"
+    }
+  }
+}
+
+resource "aws_iam_role" "doit_eks_lens_collector" {
+  count = var.ec2_cluster ? 0 : 1
+
+  name               = "doit_eks_${local.region}_${var.cluster.name}"
+  assume_role_policy = data.aws_iam_policy_document.doit_eks_lens_collector[count.index].json
+}
+
+resource "aws_iam_role_policy_attachment" "doit_eks_lens_collector" {
+  count = var.ec2_cluster ? 0 : 1
+
+  role       = aws_iam_role.doit_eks_lens_collector[count.index].name
+  policy_arn = local.s3_policy_arn
+}
+
+// Required if running EKS Lens Collector on a self-managed cluster on EC2
+resource "aws_iam_user" "doit_eks_lens_collector" {
+  count = var.ec2_cluster ? 1 : 0
+  name  = "doit_eks_lens_collector_${local.region}_${var.cluster.name}"
+  path  = "/"
+
+  force_destroy = true
+}
+
+resource "aws_iam_access_key" "doit_eks_lens_collector" {
+  count = var.ec2_cluster ? 1 : 0
+  user  = aws_iam_user.doit_eks_lens_collector[count.index].id
+}
+
+resource "aws_iam_user_policy_attachment" "iam_policy" {
+  count      = var.ec2_cluster ? 1 : 0
+  policy_arn = local.s3_policy_arn
+  user       = aws_iam_user.doit_eks_lens_collector[count.index].name
+}
+
+// Wait 60s for collector to run and start writing metrics to S3
+resource "time_sleep" "wait_60_seconds" {
+  depends_on = [kubernetes_deployment.collector]
+
+  create_duration = "60s"
+}
+
+// Register/de-register cluster on DoiT
+resource "null_resource" "doit_webhook" {
+  depends_on = [time_sleep.wait_60_seconds]
+
+  provisioner "local-exec" {
+    when       = create
+    command    = <<EOT
+    curl -X POST -H 'Content-Type: application/json' -d '{
+      \"account_id\": \"${local.account_id}\",
+      \"region\": \"${local.region}\",
+      \"cluster_name\": \"${var.cluster.name}\",
+      \"deployment_id\": \"${var.cluster.name}\"
+    }' ${local.doit_webhook_url}/terraform-validate"
+    EOT
+    on_failure = fail # in case this fails, the resources will be tainted and a re-apply will be necessary
+  }
+
+  provisioner "local-exec" {
+    when       = destroy
```

Comment:
I don't think this will work as expected, since if the cluster module is removed entirely, this resource won't be in the configuration anymore, so this won't run: https://developer.hashicorp.com/terraform/language/resources/provisioners/syntax#destroy-time-provisioners

I'm not sure how to work around this best - I was thinking we might be able to leverage a conditional [`prevent_destroy`](https://developer.hashicorp.com/terraform/tutorials/state/resource-lifecycle#prevent-resource-deletion) somehow, but not sure if it's feasible.

Reply:
Hmm, well I'm assuming that if somebody decides to get rid of this TF configuration they would also destroy it beforehand?

Reply:
Hm, would they? I'd expect the just remove the `module` block and apply? At which point, if I understand this correctly, this hook wouldn't run.

Reply:
This is possible, in which case the README gives the command to offboard... I'm not sure there's a better way to automate this.
Honestly, I would put a TTL on our side to offboard any cluster that is unresponsive for more than X days

Reply:
@hobord for your consideration